utilities/transactions/lock/range/range_tree/range_tree_lock_manager.cc

```diff
@@ -0,0 +1,572 @@
+#include "range_tree_lock_tracker.h"
+#include "rocksdb/slice.h"
+#include "rocksdb/utilities/transaction_db_mutex.h"
+#include "test_util/sync_point.h"
+#include "util/cast_util.h"
+#include "util/hash.h"
+#include "util/thread_local.h"
+#include "utilities/transactions/pessimistic_transaction_db.h"
+#include "utilities/transactions/transaction_db_mutex_impl.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+/////////////////////////////////////////////////////////////////////////////
+// RangeTreeLockManager -  A Range Lock Manager that uses PerconaFT's
+// locktree library
+/////////////////////////////////////////////////////////////////////////////
+
+RangeLockManagerHandle* NewRangeLockManager(
+    std::shared_ptr<TransactionDBMutexFactory> mutex_factory) {
+  std::shared_ptr<TransactionDBMutexFactory> use_factory;
+
+  if (mutex_factory)
+    use_factory = mutex_factory;
+  else
+    use_factory.reset(new TransactionDBMutexFactoryImpl());
+
+  return new RangeTreeLockManager(use_factory);
+}
+
+static const char SUFFIX_INFIMUM = 0x0;
+static const char SUFFIX_SUPREMUM = 0x1;
+
+void serialize_endpoint(const Endpoint& endp, std::string* buf) {
+  buf->push_back(endp.inf_suffix ? SUFFIX_SUPREMUM : SUFFIX_INFIMUM);
+  buf->append(endp.slice.data(), endp.slice.size());
+}
+
+void deserialize_endpoint(const DBT* dbt, EndpointWithString* endp) {
+  assert(dbt->size >= 1);
+  const char* dbt_data = (const char*)dbt->data;
+  char suffix = dbt_data[0];
+  assert(suffix == SUFFIX_INFIMUM || suffix == SUFFIX_SUPREMUM);
+  endp->inf_suffix = (suffix == SUFFIX_SUPREMUM);
+  endp->slice.assign(dbt_data + 1, dbt->size - 1);
+}
+
+// Get a range lock on [start_key; end_key] range
+Status RangeTreeLockManager::TryLock(PessimisticTransaction* txn,
+                                     uint32_t column_family_id,
+                                     const Endpoint& start_endp,
+                                     const Endpoint& end_endp, Env*,
+                                     bool exclusive) {
+  toku::lock_request request;
+  request.create(mutex_factory_);
+  DBT start_key_dbt, end_key_dbt;
+
+  TEST_SYNC_POINT("RangeTreeLockManager::TryRangeLock:enter");
+  std::string start_key;
+  std::string end_key;
+  serialize_endpoint(start_endp, &start_key);
+  serialize_endpoint(end_endp, &end_key);
+
+  toku_fill_dbt(&start_key_dbt, start_key.data(), start_key.size());
+  toku_fill_dbt(&end_key_dbt, end_key.data(), end_key.size());
+
+  // Use the txnid as "extra" in the lock_request. Then, KillLockWait()
+  // will be able to use kill_waiter(txn_id) to kill the wait if needed
+  // TODO: KillLockWait is gone and we are no longer using
+  // locktree::kill_waiter call. Do we need this anymore?
+  TransactionID wait_txn_id = txn->GetID();
+
+  auto lt = get_locktree_by_cfid(column_family_id);
+
+  request.set(lt, (TXNID)txn, &start_key_dbt, &end_key_dbt,
+              exclusive ? toku::lock_request::WRITE : toku::lock_request::READ,
```

Comment:
The comment mentioned that non-exclusive lock is not supported. Then what's the semantics of `READ` here?

Reply:
We don't support shared range locks, but we do support shared point locks.  

(Rationale for this: initially, locktree didn't support shared locks, and so Range Locking code didn't support it either.  Then, there was a use case for MyRocks which required that shared point locks are supported. So, limited support for shared point locks was added. Sharing point locks are much easier as one doesn't have to deal with locks on partially-overlapping ranges.)